Implement List component for Comments

dvdchr opened the pull request:
Resolves #16851
Refs #16740, #16741 
Depends on #16843

## Summary
This PR introduces the new List component to Comments list, with a feature flag. Here's a summary of the changes:
- A new feature flag, `unifiedCommentsAndNotificationsList` is added. The default behavior is `true` for **local developer and integration builds**. This is to ease testing for developers and designers, without having to make any manual changes.
- The relative date section identifier for `Comment` is copied from `Notifications+Interface`. Originally, I wanted to create a protocol (`RelativeDateSectionIdentifiable`) with default implementation, but since `Comment` and `CommentsViewController` are still in Objective-C, protocol extensions don't work. Hence the copy paste solution for now. 😢 
- Since `Comment` can pull avatar image either from URL or from gravatar email, I changed the image download approach in `ListTableViewCell` to be similar to `CommentsTableViewCell` – The `imageURL` prop is removed, and two methods are introduced instead: `configureImage(with url:)` that accepts a URL and `configureImageWithGravatarEmail`, that accepts a String.
- The table view footer in `CommentsViewController` is also updated from `CGRectZero` to a UIView with 1pt height. This is to get rid of the separator for the last row _on the last section_. Previously, the last row in the last section was still showing a separator line.

## To Test

#### Scenario 1: With disabled feature flag, Comments list should be displayed with the old design
- Go to My Site > Comments.
- Verify that the comments list is displayed with the old design.
- Verify that the comments are sectioned by individual date, instead of relative date range.

#### Scenario 2: With enabled feature flag, Comments list should be displayed with the new design
- Go back to My Site, click your Avatar, and go to App Settings.
- Select the Debug menu, and scroll down until you find `Unified Comments and Notifications List` feature flag.
- Make sure that the feature flag is **enabled**.
- Go back to My Site > Comments.
- Verify that the comments list is displayed with the new design.
- Verify that the comments are sectioned correctly.

## Regression Notes
1. Potential unintended areas of impact
Comments list not showing correctly when feature flag is turned off.

2. What I did to test those areas of impact (or what existing automated tests I relied on)
Ensured that the new and old list are displayed correctly when the feature flag is toggled.

3. What automated tests I added (or what prevented me from doing so)

PR submission checklist:

- [x] I have completed the Regression Notes.
- [x] I have considered adding unit tests for my changes.
- [x] I have considered adding accessibility improvements for my changes.
- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

dvdchr committed:
Add feature flag for unified list

dvdchr committed:
Add regular and bold text attributes for List style

dvdchr committed:
Add ListTableViewCell+Comments for Comments-specific cell configuration

dvdchr committed:
Add capability to download avatar image from Gravatar email

dvdchr committed:
Add objc-accessible reuseIdentifier and estimatedRowHeight

dvdchr committed:
Implement List component in Comments (with feature flag)

dvdchr committed:
Update footer view to use a 1pt clear UIView

With the previous implementation, the last row in the last section was still showing the separator line. This change correctly "hides" the separator lines for last rows in all sections.

dvdchr committed:
Add relative date section identifier for Comment

dvdchr committed:
Section comments by relative date (following Notifications)

dvdchr committed:
Fix minor objc code style

ScoutHarris commented on WordPress/Classes/ViewRelated/Comments/CommentsViewController.m:
```diff
@@ -131,12 +136,19 @@ - (void)configureTableView
     UINib *nibInstance = [UINib nibWithNibName:nibName bundle:[NSBundle mainBundle]];
     [self.tableView registerNib:nibInstance forCellReuseIdentifier:CommentsTableViewCell.reuseIdentifier];
+
+    // Register unified list components
```
Can the cell registrations be based on `usesUnifiedList`? At the very least, it will be clear later when the feature flag is removed what needs to be removed/kept.

dvdchr replied:
That makes sense! Will adjust this soon.

ScoutHarris commented on WordPress/Classes/ViewRelated/Comments/CommentsViewController.m:
```diff
@@ -131,12 +136,19 @@ - (void)configureTableView
 {
     // Hide the cellSeparators when the table is empty
-    self.tableView.tableFooterView = [[UIView alloc] initWithFrame:CGRectZero];
+    self.tableView.tableFooterView = [[UIView alloc] initWithFrame:CGRectMake(0, 0, self.tableView.frame.size.width, 1)];
```
👍

ScoutHarris commented on WordPress/Classes/ViewRelated/Comments/CommentsViewController.m:
```diff
@@ -165,8 +206,15 @@ - (CGFloat)tableView:(UITableView *)tableView heightForRowAtIndexPath:(NSIndexPa
 - (UITableViewCell *)tableView:(UITableView *)tableView cellForRowAtIndexPath:(NSIndexPath *)indexPath
 {
+    if ([self usesUnifiedList]) {
+        ListTableViewCell *listCell = (ListTableViewCell *)[tableView dequeueReusableCellWithIdentifier:ListTableViewCell.reuseIdentifier
```
There should probably be a check here to ensure `listCell` exists (similar to the existing `if (!cell)` logic).

dvdchr replied:
Since I'm using `dequeueReusableCellWithIdentifier:forIndexPath:`, the method guarantees that it returns a non-null instance of the cell, as long as it's properly registered:

```objc
// newer dequeue method guarantees a cell is returned and resized properly, assuming identifier is registered
- (__kindof UITableViewCell *)dequeueReusableCellWithIdentifier:(NSString *)identifier forIndexPath:(NSIndexPath *)indexPath API_AVAILABLE(ios(6.0));
```

I think it's unnecessary to add additional `if (!cell)` checks, but I'd like to double check this again with you 😄 . If you think we still need the check, I'll add it! 🙂

ScoutHarris replied:
> as long as it's properly registered

Ah, this part I did not know. Reread the documentation for that method, and now I do. 😄 I was distracted by the old code. Sorry for the bother!

ScoutHarris commented on WordPress/Classes/ViewRelated/Views/List/ListTableHeaderView.swift:
```diff
@@ -12,6 +12,9 @@ class ListTableHeaderView: UITableViewHeaderFooterView, NibReusable {
 
     // MARK: Properties
 
+    @objc static let reuseIdentifier = defaultReuseID
```
Nitpick/request - can a comment be added here indicating `reuseIdentifier` can be removed when the VCs are converted to Swift? It'd be nice to _eventually_ use `defaultReuseID` directly.

dvdchr replied:
Noted! Will do.

ScoutHarris commented on WordPress/Classes/ViewRelated/Views/List/ListTableViewCell.swift:
```diff
@@ -14,6 +14,9 @@ class ListTableViewCell: UITableViewCell, NibReusable {
 
     // MARK: Properties
 
+    @objc static let reuseIdentifier = defaultReuseID
```
Same nitpick/request as above - can a comment be added here indicating `reuseIdentifier` can be removed when the VCs are converted to Swift? It'd be nice to _eventually_ use `defaultReuseID` directly.

dvdchr replied:
Yup! Agreed. 😄  Will add docs here too.

dvdchr committed:
Register comment cells based on feature flag

dvdchr committed:
Add docs for List components on NibReusable objc support

ScoutHarris reviewed (approved):
Thanks for the changes! The list looks great!

I just spotted one thing - when showing the new UI, these appear in the logs:
```
2021-07-14 11:39:50.309944-0600 WordPress[17754:193539] [TableView] Changing the background color of UITableViewHeaderFooterView is not supported. Use the background view configuration instead.
2021-07-14 11:39:50.315172-0600 WordPress[17754:193539] [TableView] Changing the background color of UITableViewHeaderFooterView is not supported. Use the background view configuration instead.
```

Since that is unrelated to this PR, it can be fixed separately. So GTG here.

:shipit:

dvdchr commented:
On second thought, I think I'll set the feature flag to `false` right now, so folks don't suddenly experience the change. I'll re-enable to local developer + integration builds again in the PR for Notifications.

> I just spotted one thing - when showing the new UI, these appear in the logs:

Good catch, and what a pair of 🦅  👀  ! ~I'll try to sneak in the fix here.~ Done! 

Plus, I added checks for table header view instantiation, since the method is nullable.

dvdchr committed:
Fix warnings and add minor safety changes

- Fixed table header view background color warning.
- Added nullability checks for table header view instantiation.

dvdchr committed:
Fix minor comment typo

dvdchr committed:
Set the feature flag to false
